doc/generating-code.mld

```diff
@@ -0,0 +1,357 @@
+{%html: <div style="display: flex; justify-content:space-between"><div>%}{{!"writing-ppxs"}< Writing PPXs}{%html: </div><div>%}{{!"matching-code"}Destructing AST nodes >}{%html: </div></div>%}
+
+{0 Generating AST Nodes}
+
+The rewriter's core is a function that outputs code in the form of an AST.
+However, there are some issues with generating AST values when using the
+constructors directly:
+
+- The type is {{!Ppxlib.Parsetree}pretty verbose}, with many fields
+  rarely used.
+- The AST type might change at a version bump. In this case, the types used in
+the PPX would become incompatible with the types of the new OCaml version.
+
+The second point is important: since [ppxlib] {{!page-driver.compat_mult_ver}translates} the AST to
+the newest OCaml AST available before rewriting, your PPX would not
+only become incompatible with the new OCaml version, but also with all [ppxlib]
+versions released after the new AST type is introduced.
+
+For this reason, [ppxlib] provides abstractions over the OCaml AST, with a focus
+on usability and stability.
+
+{1 The Different Options}
+
+The two main options are:
+
+- {{!Ppxlib.Ast_builder}[Ast_builder]},
+- {!Ppxlib_metaquot}.
+
+{{!Ppxlib.Ast_builder}[Ast_builder]} provides an API to generate AST nodes for the latest OCaml
+version in a backward-compatible way. {!Ppxlib_metaquot} is different: it is a PPX that
+lets you generate OCaml AST nodes by writing OCaml code, using quoting and
+splicing.
+
+Using {!Ppxlib_metaquot} requires less knowledge of the OCaml AST than
+{{!Ppxlib.Ast_builder}[Ast_builder]} as it only uses natural OCaml syntax;
+however, it is less general because it only allows users to generate few different nodes such as
+structure items, expressions, patterns, etc., but it is not possible to generate a
+value of type {{!Ppxlib.Parsetree.row_field_desc}[row_field_desc]}!
```

Comment:
I would not say that it lacks the ability to generate nodes dynamically: antiquotation is a solution for this. The problem is that anti-quotation is a dynamic solution on its own, _but is limited to a few nodes_, so both problems are actually kind of the same!

For instance, the "rec" keyword cannot be generated dynamically: 
```
let is_recursive = ...
let _ =
  [%str let [%a is_recursive] name = ... ]
```
precisely because there is no way to generate just a "rec" value and there is no way to antiquote only the "rec" value.

I'm happy to rephrase, but I'm reluctant to say that "it lacks the ability to generate nodes dynamically" since it would be falsified by the next section!

Reply:
> I would not say that it lacks the ability to generate nodes dynamically

What I'm trying to transmit is that `metaquot` _on its own_ isn't able to generate nodes dynamically. For a dynamic node, you need anti-quotation + `Ast_builder` (or another means to build the node).

> I'm happy to rephrase, but I'm reluctant to say that "it lacks the ability to generate nodes dynamically" since it would be falsified by the next section!

You've skipped the "on its own" here :) In the next section, the value you call `dynamic_node` is constructed via `Ast_builder`. So the example doesn't use `metaquot` on its own: the dynamic part is constructed via `Ast_builder` + `mataquot` anti-quotation.

I've tried to make that clear with my last sentence "A typical workflow is to use `metaquot` for the constant skeleton of the node, and to use the `metaquot` anti-quotation workflow (see below) together with `Ast_builder` to fill in the dynamic parts.".

I clearly haven't made clear what I wanted to say, though. Do you agree that it would be nice to add something around this? If not, that's ok with me. If so, let me know if you want me to write a more detailed suggestion!

Reply:
Ok, I think I understand. I was confused by the fact that it is _sometimes_ possible to use metaquot on its own to generate dynamic node:
```
let e = if cond then [%expr 1+1] else [%expr 2+2] in
[%expr 1 + [%e e]
```
However, it is very restricted in what it can generate, and from what data: for instance, it is not possible to create an expression containing a string, given the string itself. Is it what you meant?
Anyway, I agree that it would be nice to add something about this!

```suggestion
however, it's more restrictive than `Ast_builder` for two reasons: first, it's less flexible, since on its own it lacks the ability to generate nodes dynamically from other kind of data: e.g. it's not possible to build an expression containing a string, given the string as input. Second, it's less general because it only allows users to generate few different nodes such as
structure items, expressions, patterns, etc., but it is not possible to generate a
value of type {{!Ppxlib.Parsetree.row_field_desc}[row_field_desc]}! A typical workflow is to use `metaquot` for the constant skeleton of the node, and to use the `metaquot` anti-quotation workflow (see below) together with `Ast_builder` to fill in the dynamic parts.
```